### PR TITLE
Add caching for route urls to allow for more cache collisions

### PR DESCRIPTION
**Pull request by igorwulff**
This change will improve route url retrieval in combination with different variants of a url with query params and ensure query params are not taking into account when generating a cached entry. 

On a PLP we have cases where this method is called 300-900 times depending if there is block_cache or not: For a fully cached page excluding FPC, this is called 327 times and takes 31ms and with patch 12ms. For a fully cached page excluding FPC and Block_cache, this is called 836 times and takes 38ms and with patch 12.5ms.

The gains can be much bigger depending on the situation. For another client we have seen bigger gains where more variable urls within certain facets were given with query params, resulting in a bigger impact.

A different solution might be proposed of course, but I think the potential gain should be looked at.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Looks like the changes causing test failures. Please review and fix them

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Static Tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
Moving back this PR to the Changes requested

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE

**Comment by engcom-Hotel** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests EE, Magento Health Index, Sample Data Tests B2B, Sample Data Tests CE, Sample Data Tests EE, Semantic Version Checker, Static Tests, Unit Tests, WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests EE

File: lib/internal/Magento/Framework/Url.php [hunk 1]
```diff
@@ -1008,6 +1014,7 @@ public function getRebuiltUrl($url)
      * @param string $value
      * @return string
      * @deprecated 101.0.0
+     * @see https://github.com/magento-commerce/magento2ce/commit/6729b6e01368248abc33300208eb292c95050203
```
Comment (hostep) [inline lib/internal/Magento/Framework/Url.php]: @glo24157: no idea if this url is the best way to explain the deprecation, but you should at least link to a publicly available url, so this one is better `https://github.com/magento/magento2/commit/6729b6e01368248abc33300208eb292c95050203`

Linking to private repositories is definitely not helpful.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
Look into the test failures. The failures seem to falky. Please find below the screenshot for reference:

### Integration Tests Failures
testUnassignProductsWithSameSku
testUnassignProductWithNonExistentSku

testAfterGetQuery
testAssignProductsWithSameSku
testGetAssignedProductsIds
testGetAssignedProductsSkus

### Functional Test Failures

Regarding functional test failures, the failed test is not due to the PR changes. The tests are failing on the fresh 2.4-develop instance as well. Hence it should be flaky.

Moving this PR to pending review status.

Thanks

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run Integration Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by ihor-sviziev** [comment]
hi @glo24157, I see you changed the implementation, which probably might significantly reduce performance improvement. Did you compare performance diff between these solutions?

**Comment by engcom-Hotel** [comment]
Hello @igorwulff,

Request you to please resolve the conflicts so that we can move further with this PR.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @igorwulff,

We are unable to proceed with this PR as there are some conflicts. Hence we are closing this PR for now.

Let us know when you are ready to work on it again, and we will be happy to re-open it. Thank you for your contribution!

Thanks

**Pull request closed without merging**